Fix hang in file watcher

slluis opened the pull request:
Enabling events has a sync wait on a flush of the IO event queue, and
that may take many seconds to complete if there is a lot of IO. To
avoid locking on this call, we now enable the events on a background
thread.

Fixes VSTS #952405 - Visual Studio Mac 2019 Freezes on Switch Branch
Fixes VSTS #980470 - Deadlock in FileWatcherService.WatchDirectories

slluis committed:
Fix hang in file watcher

Enabling events has a sync wait on a flush of the IO event queue, and
that may take many seconds to complete if there is a lot of IO. To
avoid locking on this call, we now enable the events on a background
thread.

Fixes VSTS #952405 - Visual Studio Mac 2019 Freezes on Switch Branch
Fixes VSTS #980470 - Deadlock in FileWatcherService.WatchDirectories

slluis commented:
@monojenkins backport release-8.3

slluis commented:
@monojenkins backport release-8.3

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs:
```diff
@@ -363,18 +367,71 @@ public FileWatcherWrapper (FilePath path, PathTreeNode rootNode, ReaderWriterLoc
+						// Enabling events has a sync wait on a flush of the IO event queue, and that may take many
+						// seconds if there is a lot of IO. To avoid locking on this call, we enable the events
+						// on a background thread.
+						Task.Run (EnableEvents).Ignore ();
```
Maybe we return this task so the service can return the task that actually waits for the watcher to be enabled.

Therzok replied:
UpdateWatchers would then return a Task and UpdateWatchersAsync would return the Task.WhenAll for all of these.

Therzok replied:
Otherwise we might see transient failures in VSTS because the directories aren't watched because of high IO.

mrward replied:
Yeah. I wonder if this will affect some tests. Some tests assume the watcher is created when file watcher service methods return which might not happen now. Have to see.

https://github.com/mono/monodevelop/blob/dea5016c423bcc2f827a557d33c32cc2e5e75e22/main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs#L309

Therzok reviewed (changes requested):
Marking as requesting changes. We need to handle the watcher being enabled by handling the task.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs:
```diff
@@ -163,29 +163,32 @@ static Task UpdateWatchersAsync ()
 			CancellationToken token = cancellationTokenSource.Token;
 
-			return Task.Run (() => UpdateWatchers (token));
+			return Task.Run (async () => await UpdateWatchers (token));
```
```suggestion
			return Task.Run (() => UpdateWatchers (token)).Unwrap ();
```

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs:
```diff
@@ -212,24 +215,27 @@ static void UpdateWatchers (CancellationToken token)
+								watchers.Remove (path);
+							}
+						}
+					}, TaskScheduler.Default);
```
```suggestion
					}, CancellationToken.None, TaskContinuationOptions.OnlyOnFaulted, TaskScheduler.Default);
```

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs:
```diff
@@ -212,24 +215,27 @@ static void UpdateWatchers (CancellationToken token)
-						watchers.Remove (path);
-					}
+					var task = watcher.EnableRaisingEventsAsync ().ContinueWith (t => {
+						if (t.IsFaulted && t.Exception.InnerException is UnauthorizedAccessException e) {
```
Not needed anymore with TaskContinuationOptions.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs:
```diff
@@ -363,18 +376,88 @@ public FileWatcherWrapper (FilePath path, PathTreeNode rootNode, ReaderWriterLoc
+			get { return eventsEnabled; }
+		}
+
+		public void DisableRisingEvents ()
```
```suggestion
		public void DisableRaisingEvents ()
```

and other places.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs:
```diff
@@ -363,18 +376,88 @@ public FileWatcherWrapper (FilePath path, PathTreeNode rootNode, ReaderWriterLoc
+			return Task.CompletedTask;
 		}
 
+		void EnableEvents (Task t)
```
t isn't used

slluis replied:
it is necessary to avoid a lambda in ContinueWith

Therzok replied:
I'll be pedantic a bit, but roslyn codegen sucks.

This reuses a cached `Action<Task>`
```csharp
ContinueWith (t => EnableEvents (t))
```

This always allocates an `Action<Task>`.
```csharp
ContinueWith (EnableEvents);
```

Therzok replied:
Ah, nevermind! This is not a lambda without captures.

Therzok reviewed (approved):
Looks good!

slluis commented:
@monojenkins backport release-8.3

slluis commented:
@monojenkins backport release-8.3

slluis committed:
Fix FSW update logic

When updating watchers, if a watcher already exists ensure it is fully
initialized before leaving the Update method.

slluis commented:
@monojenkins backport release-8.3

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs:
```diff
@@ -209,8 +209,9 @@ static Task UpdateWatchers (CancellationToken token)
-					if (watchers.ContainsKey (path)) {
+					// Don't modify a watcher that already exists, but ensure it is initialized
+					if (watchers.TryGetValue (path, out var existingWatcher)) {
+						tasks.Add (existingWatcher.EnableRaisingEventsAsync ());
```
:+1:
